sympy/core/evalf.py

```diff
@@ -1558,6 +1559,15 @@ class EvalfMixin:
 
     __slots__ = ()  # type: tTuple[str, ...]
 
+    @overload
+    def evalf(self: Expr, n: int = 15, subs: dict[Basic, Basic | float] | None = None, # type: ignore
+              maxn: int = 100, chop: bool = False, strict: bool  = False,
+              quad: str | None = None, verbose: bool = False) -> Expr: ...
+    @overload
+    def evalf(self: Basic, n: int = 15, subs: dict[Basic, Basic | float] | None = None, # type: ignore
```

Comment:
Basic does not have the `evalf` though. 
```python
>>> from sympy import Basic
>>> Basic(1, 2, 3).evalf()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Basic' object has no attribute 'evalf'
```

I may have other suggestion that `Expr` and a handful of places where EvalfMixin is used could override the method trivially with the different specified types.

```python
    def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None, verbose=False) -> Expr:
        return super().evalf(n, subs, maxn, chop, strict, quad, verbose) # type: ignore
```

Reply:
> Basic does not have the evalf though.

Anyway, I see that it is no problem because mypy and pyright correctly reports errors for `Basic` instances with no evalf.
However, I'd like to consider overriding to be better choice later because `Set, Relational, GeometryEntity` for example, are bounded themselves as `Set, Relational, GeometryEntity` for evalf.